### ISSUE TITLE
Support to update a container in the StatefulSet pods without recreating the pods during rolling update. 

### ISSUE THREAD

**Issue by feellifexp** [kubernetes/kubernetes#61142]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature

**What happened**:
Currently when rolling update a StatefulSet to update containers for, binary release, the controller will delete the current pod then create a corresponding new pod, one by one. 

The problems with this approach are:
- The newly created pod may not be scheduled to the same node as the previous pod. This could have impact to some services in favor of locality, for example HBase.
- Recreating the pod is more expensive than just restarting the container within the pod.

**What you expected to happen**:
- For the container changes that are safe to apply, only the changed container is restarted to pick up the change. The pod is kept untouched.

**Anything else we need to know?**:
It seems to me the most challenging parts to implement this are:
- How to determine wether the container changes are safe so that they can be updated by restarting the container in place?
- How to check the healthiness of the pod N after its containers are updated before updating the containers in pod N-1.

Would appreciate any feedback on whether this request is reasonable.

Thanks!

**Comment by s4rd1nh4** [comment]
@pwittrock Help please

**Comment by feellifexp** [comment]
Does this request sound reasonable? Is there plan to support this? I can volunteer to spend time on the  investigation/implementation if it's in the roadmap of sig-apps, and can be supported.

Thanks!

**Comment by bgrant0607** [comment]
@feellifexp The short answer is theoretically yes, in practice it would be hard. Almost nothing in pods is mutable currently.

See also #9043.

**Comment by maguro** [comment]
TL;DR maybe enhancements to affinity/taints would better suit your neeeds.

For me, this feature would complicate things.  I like to think of the pod as an atomic unit.  Doing so makes things easier to reason about. 

As for the concerns that are the impetus for this issue, assuming that affinity/taints as they are now are not adequate to your purposes, my thinking is that being concerned about what specific node your pod is assigned to seems to be a bit antithetical to what k8s is all about.  But If there was some valid reason for doing so, then wouldn’t it be better to have something akin to affinity/taints which would make your need explicit?

Next, thinking upon a pod that is too burdensome to restart.  Without any details to go on,  I have to assume that there are some side cars that are very sluggish. If that’s the case, wouldn’t it be better to move them to another pod? It seems that again having something akin to affinity/taints would help here as well.

Another thing to consider is facilitating the participation of operators and the like.  At the pod level, there are a number of mechanisms, al beit a bit unbalanced and wonky, that facilitate the participation of operators in a pod’s lifecylcle.  Maybe you don’t care about it but, for this to be framework feature, some serious consideration need to be given on the lifecycle/operator front.

Finally, keeping in mind that I don’t have the same specific problems as you, I don’t worry too much about how long it takes to restart a pod or update a cluster, within reason.   The size of the clusters, the load they bear, and the constant flow of releases means that cluster is always in a state of flux, not unlike the maintenance of the Golden Gate Bridge; it took me a while to get used to. For me, what’s more important is clear and simple behavior that makes reasoning about what’s going on tractable.
